### PR TITLE
Change merge type check from dict to Mapping

### LINKED ISSUE #468: Toolz.merge fails when provided one non-dict mapping
Here is a minimal reproducible example:

```python
In [1]: import toolz, collections

In [2]: class Foo(collections.Mapping):
   ...:     def __init__(self, d):
   ...:         self.d = d
   ...:     def __iter__(self):
   ...:         return iter(self.d)
   ...:     def __getitem__(self, key):
   ...:         return self.d[key]
   ...:     def __len__(self):
   ...:         return len(self.d)
   ...:

In [3]: d = Foo({1: 1})

In [4]: toolz.merge(d)
```
```python-traceback
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-e5a2d80dd87f> in <module>
----> 1 toolz.merge(d)

~/miniconda/envs/dev/lib/python3.7/site-packages/toolz/dicttoolz.py in merge(*dicts, **kwargs)
     36     rv = factory()
     37     for d in dicts:
---> 38         rv.update(d)
     39     return rv
     40

TypeError: 'int' object is not iterable
```

I think that we might want to change the `dict` typecheck here to `Mapping`

### PR DESCRIPTION
**Pull request by mrocklin**
Fixes #468

### CODE REVIEW

**Comment by tsanikgr** [comment]
Twitter brought me here!

File: toolz/dicttoolz.py [hunk 1]
```diff
@@ -29,7 +30,7 @@ def merge(*dicts, **kwargs):
     See Also:
         merge_with
     """
-    if len(dicts) == 1 and not isinstance(dicts[0], dict):
+    if len(dicts) == 1 and not isinstance(dicts[0], collections.Mapping):
```
Comment (jcrist) [inline toolz/dicttoolz.py]: Mapping should be imported from `collections.abc`, not doing this will fail in Python 3.8 and warn otherwise.

Comment (mrocklin) [reply toolz/dicttoolz.py]: Thanks!  Done.

**Comment by eriknw** [comment]
Nice, and good find!  Can we update `merge_with` while we're at it?

**Comment by eriknw** [comment]
Thanks for the fix @mrocklin, it's good to have you around!  I'm surprised this edge case has remained for so long.  I'll try to get a new release out with Python 3.8 within a week or two (my "real life" remains very busy, but I should be able to take care of this during working hours).

**Pull request merged**